presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestRemoveIdentityCasts.java

```diff
@@ -0,0 +1,127 @@
+import com.facebook.presto.tpch.TpchConnectorFactory;
+import com.google.common.collect.ImmutableMap;
+import org.intellij.lang.annotations.Language;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.anyTree;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+
+public class TestRemoveIdentityCasts
+{
+    private final LocalQueryRunner queryRunner;
+
+    public TestRemoveIdentityCasts()
+    {
+        this.queryRunner = new LocalQueryRunner(testSessionBuilder()
+                .setCatalog("local")
+                .setSchema("tiny")
+                .build());
+
+        queryRunner.createCatalog(queryRunner.getDefaultSession().getCatalog().get(),
+                new TpchConnectorFactory(queryRunner.getNodeManager(), 1),
+                ImmutableMap.<String, String>of());
+    }
+
+    @Test
+    public void testRemoveIdentityCastUnparameterizedColumnData()
+    {
+        @Language("SQL") String sql = "SELECT CAST(nationkey AS bigint) FROM nation";
+        PlanMatchPattern pattern = anyTree(
+                project(anyTree()).withAssignment("Cast")
+        );
+        assertNotPlan(sql, pattern);
+    }
+
+    @Test
+    public void testKeepNonIdentityCastUnparameterizedColumnData()
+    {
+        @Language("SQL") String sql = "SELECT CAST(nationkey AS smallint) FROM nation";
+        PlanMatchPattern pattern = anyTree(
+                project(anyTree()).withAssignment("Cast")
+        );
+        assertPlan(sql, pattern);
+    }
+
+    @Test
+    public void testRemoveIdentityCastParameterizedColumnData()
+    {
+        @Language("SQL") String sql = "SELECT CAST(name AS varchar(25)) FROM nation";
+        PlanMatchPattern pattern = anyTree(
+                project(anyTree()).withAssignment("Cast")
+        );
+        assertNotPlan(sql, pattern);
+    }
+
+    @Test
+    public void testKeepNonIdentityCastParameterizedColumnData()
+    {
+        @Language("SQL") String sql = "SELECT CAST(name AS varchar(30)) FROM nation";
+        PlanMatchPattern pattern = anyTree(
+                project(anyTree()).withAssignment("Cast")
+        );
+        assertPlan(sql, pattern);
+    }
+
+    @Test
+    public void testKeepNonIdentityCastDefaultParameterizedColumnData()
+    {
+        @Language("SQL") String sql = "SELECT CAST(name AS varchar) FROM nation";
+        PlanMatchPattern pattern = anyTree(
+                project(anyTree()).withAssignment("Cast")
+        );
+        assertPlan(sql, pattern);
+    }
+
+    @Test
+    public void testRemoveIdentityCastImplicitBigintData()
+    {
+        @Language("SQL") String sql = "SELECT CAST(bigint '5' AS bigint)";
+        PlanMatchPattern pattern = anyTree(
+                project(anyTree()).withAssignment("Cast")
+        );
+        assertNotPlan(sql, pattern);
```

Comment:
From what I understand:

With `assertNotPlan()`, it's saying that there exists no `ProjectNode` with an expression that is of type `Cast`. This can match against a lot of different plans that it probably shouldn't match with. I could add additional assertions to say that the plan has a `ProjectNode` to make sure it's not just a bogus plan.

With an extension of `anyNot()`, it would say that there exists a `ProjectNode` that does not have a `Cast` expression. I don't think this is as helpful as the first case, at least for what I'm trying to test. If there are a couple of `ProjectNode`s where one has a `Cast` and the other doesn't, then it will still match. I don't think there's an easy way to work around that scenario.